File: Bit Manipulation/Java/XOR Queries of a Subarray.java
// Problem link: https://leetcode.com/problems/xor-queries-of-a-subarray/description
// Time: O(n)
// Space: O(n)
class Solution {
    public int[] xorQueries(int[] arr, int[][] queries) {
        int size = arr.length;
        int[] store = new int[size + 1];
        int[] result = new int[queries.length];

        for(int idx = 0; idx < size; idx++) {
            store[idx + 1] = store[idx] ^ arr[idx];
        }

        for(int idx = 0; idx < queries.length; idx++) {
            result[idx] = store[queries[idx][1] + 1] ^ store[queries[idx][0]]; 
        }

        return result;
    }
}

File: Bit Manipulation/Python/XOR Queries of a Subarray.py
# Problem link: https://leetcode.com/problems/xor-queries-of-a-subarray/description
# Time: O(n)
# Space: O(n)
class Solution:
    def xorQueries(self, arr: List[int], queries: List[List[int]]) -> List[int]:
        store, result = [0], []
        
        for num in arr:
            store.append(store[-1] ^ num)

        for x, y in queries:
            result.append(store[y + 1] ^ store[x])
        
        return result
